---
title: "Wright-Fisher model and Hardy-Weinberg principle simulation"
author: "Marc-Olivier Beausoleil"
date: "`r Sys.Date()`"
output:
  html_document:
    self_contained: no
bibliography: references/wf_model_simul.bib
csl: styles/evolution.csl
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align = 'center')
```

```{r prepare_R, echo = FALSE, warning=FALSE, results='hide'}
suppressMessages(source("../scripts/0.initialize.R"))

# load the functions
source("../scripts/WF_genetic_drift_simulation.R")
```

# Introduction 

Population genetics is a field of study interested in the precise mechanisms describing evolutionary processes [@hartl_principles_2007]. Of the 'big five' evolutionary processes (mutation, natural selection, gene flow, mating patterns, genetic drift), genetic drift can be expressed with the Wright-Fisher model which 'isolates' only the effect of genetic drift. In this report, we will explore the Wright-Fisher model showing genetic drift and another important basic principle in population genetics: what happens to genotype frequencies when there is no evolution (Hardy-Weinberg principle)!  

# The Wright-Fisher Model

Here is an example of simulation that was made in `R` to show genetic drift without natural selection. The same script, or function, can be used to generate patterns like these, but by varying the parameters (population sizes or starting alleles). 

Each plot below were generated with 100 populations with an allele frequency $p$ of 50% (so $q = 0.5$) as an initial condition. The only thing that varies from the 4 plots is the number of individuals within each population (note that the random seed is always the same). We can see now see with the simulation that: 

  * No **particular** allele will be fixated (unbiased);
  * the **number of individuals** seems an important factor in the fixation of alleles. The fluctuations in allele frequency are larger in smaller populations;
  * some genetic variation can be **lost**;
  * **makes populations different**, even if the pop are identical at the beginning;
  * **Fixation** of an allele without natural selection.

```{r wf_simulation_varying_n}
par(mfrow = c(2,2), mar = c(4,4,3,1))
nsim = 100 # number of replicate populations
ngen = 500 # number of generations 
pstart = 0.5 # Allele frequency at start 
seed = 1245
wf.simulation.n.2 = wf.drift(N = 2,       p = pstart, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.20 = wf.drift(N = 20,     p = pstart, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.200 = wf.drift(N = 200,   p = pstart, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.2000 = wf.drift(N = 2000, p = pstart, N.gen = ngen, N.sim = nsim, seed = seed)
```

```{r fixation_var_n, echo=FALSE, results='hide'}
fix.rate.2 = fixation(wf.simulation.n.2)$fixation.rate
fix.rate.20 = fixation(wf.simulation.n.20)$fixation.rate
fix.rate.200 = fixation(wf.simulation.n.200)$fixation.rate
fix.rate.2000 = fixation(wf.simulation.n.2000)$fixation.rate

mfix.2 = fixation(wf.simulation.n.2)$mean.fix
mfix.20 = fixation(wf.simulation.n.20)$mean.fix
mfix.200 = fixation(wf.simulation.n.200)$mean.fix
mfix.2000 = fixation(wf.simulation.n.2000)$mean.fix
```

The rate at which the alleles are fixed in these population is `r fix.rate.2` %, `r fix.rate.20` %, `r fix.rate.200` %, `r fix.rate.2000` % for the 2, 20, 200, 2000 individuals respectively. 

The mean generation at which there is fixation is `r round(mfix.2,2)`, `r round(mfix.20,2)`, `r round(mfix.200,2)`, `r round(mfix.2000,2)` for the 2, 20, 200, 2000 individuals respectively (if NaN, it means that it was never fixed). 

We can play the same game but keeping the number of individuals constant (here 20), but changing the allele frequency at the start of the simulation. 

```{r wf_simulation_n_low_varying_p}
par(mfrow = c(2,2), mar = c(4,4,3,1))
nid = 20
wf.simulation.n.20.p.1 = wf.drift(N = nid, p = .1, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.20.p.3 = wf.drift(N = nid, p = .3, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.20.p.5 = wf.drift(N = nid, p = .5, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.20.p.9 = wf.drift(N = nid, p = .9, N.gen = ngen, N.sim = nsim, seed = seed)
```

```{r fixation_n_low_var_p, echo=FALSE, results='hide'}
fix.rate.20.p.1 = fixation(wf.simulation.n.20.p.1)$fixation.rate
fix.rate.20.p.3 = fixation(wf.simulation.n.20.p.3)$fixation.rate
fix.rate.20.p.5 = fixation(wf.simulation.n.20.p.5)$fixation.rate
fix.rate.20.p.9 = fixation(wf.simulation.n.20.p.9)$fixation.rate

mfix.20.p.1 = fixation(wf.simulation.n.20.p.1)$mean.fix
mfix.20.p.3 = fixation(wf.simulation.n.20.p.3)$mean.fix
mfix.20.p.5 = fixation(wf.simulation.n.20.p.5)$mean.fix
mfix.20.p.9 = fixation(wf.simulation.n.20.p.9)$mean.fix
```

The rate at which the alleles are fixed in these population is `r fix.rate.20.p.1` %, `r fix.rate.20.p.3` %, `r fix.rate.20.p.5` %, `r fix.rate.20.p.9` % for 20 individuals, but an starting allele frequency $p$ of 0.1, 0.3, 0.5, and 0.9 respectively. 

The mean generation at which there is fixation is `r round(mfix.20.p.1,2)`, `r round(mfix.20.p.3,2)`, `r round(mfix.20.p.5,2)`, `r round(mfix.20.p.9,2)` for 20 individuals, but an starting allele frequency $p$ of 0.1, 0.3, 0.5, and 0.9 respectively (if NaN, it means that it was never fixed). 


This is the same as the previous set of plots, but the number of individuals was increased to 1000. 

```{r wf_simulation_n_hgh_varying_p}
par(mfrow = c(2,2), mar = c(4,4,3,1))
nid = 1e4
wf.simulation.n.1e4.p.1 = wf.drift(N = nid, p = .1, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.1e4.p.3 = wf.drift(N = nid, p = .3, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.1e4.p.5 = wf.drift(N = nid, p = .5, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.1e4.p.9 = wf.drift(N = nid, p = .9, N.gen = ngen, N.sim = nsim, seed = seed)
```

```{r fixation_n_hgh_var_p, echo=FALSE, results='hide'}
fix.rate.1e4.p.1 = fixation(wf.simulation.n.1e4.p.1)$fixation.rate
fix.rate.1e4.p.3 = fixation(wf.simulation.n.1e4.p.3)$fixation.rate
fix.rate.1e4.p.5 = fixation(wf.simulation.n.1e4.p.5)$fixation.rate
fix.rate.1e4.p.9 = fixation(wf.simulation.n.1e4.p.9)$fixation.rate

mfix.1e4.p.1 = fixation(wf.simulation.n.1e4.p.1)$mean.fix
mfix.1e4.p.3 = fixation(wf.simulation.n.1e4.p.3)$mean.fix
mfix.1e4.p.5 = fixation(wf.simulation.n.1e4.p.5)$mean.fix
mfix.1e4.p.9 = fixation(wf.simulation.n.1e4.p.9)$mean.fix
```

The rate at which the alleles are fixed in these population is `r fix.rate.1e4.p.1` %, `r fix.rate.1e4.p.3` %, `r fix.rate.1e4.p.5` %, `r fix.rate.1e4.p.9` % for 10,000 individuals, but an starting allele frequency $p$ of 0.1, 0.3, 0.5, and 0.9 respectively. 

The mean generation at which there is fixation is `r round(mfix.1e4.p.1,2)`, `r round(mfix.1e4.p.3,2)`, `r round(mfix.1e4.p.5,2)`, `r round(mfix.1e4.p.9,2)` for 10,000 individuals, but an starting allele frequency $p$ of 0.1, 0.3, 0.5, and 0.9 respectively (if NaN, it means that it was never fixed). 


# Hardy-Weinberg principle

Genotype frequencies is important measure in populations to understand if any evolutionary mechanism it at play [@hahn_molecular_2018]. A simple mathematical model was developed in 1908 to infer the genotype frequencies based on the allele frequency in a population [@hardy_mendelian_1908; @weinberg_uber_1908].

In a nutshell, here is Hardy-Weinberg's principle: 

$$ 1 = p + q \\ 
1^2 = (p + q)^2 \\ 
1 = (p + q)(p + q) \\
1 = p^2 + 2pq + q^2$$

Where $p$ is is $A$ allele frequency and $q$ is the $a$ allele frequency. The frequency of genotypes $AA$, $Aa$ and $aa$ are given by $p^2$, $2pq$, and $q^2$ respectively. 

```{r hardy_weinber_principle, fig.width=7, fig.height=5}
par(mfrow = c(1,1), mar = c(4,4,3,1))

# Make a sequence of p allele frequencies 
p.seq = seq(1,0,by = -.01)

# Calculate expected genotype frequencies
dat.hw = t(mapply(h.p.2pq.q, p = p.seq)) 

matplot(x = p.seq, 
        y = dat.hw, 
        type = "l",
        main = paste0("Hardy-Weinberg principle"),            # Title 
        xlab = paste0("Frequency of p (A) allele (or q (a) if 1-p)"), # x-axis lab title
          ylab = "Genotype frequency",                        # y-axis lab title
        ylim = c(0,1),
        lwd = 4,
        lty = 1,
        col = c("black","blue","red"))             
legend(x = .5,
       y = 1, 
       legend = c("AA","Aa","aa"), 
       col = c("black","blue","red"), 
       lty = 1, 
       lwd = 4, 
       title = "Genotypes")
```


# Saving the data

Here the data from all simulations are exported for future reference. 

```{r export_data}
path.n = "../data/WF_model/sim.n"
dir.create(path.n, showWarnings = FALSE)
write.csv(wf.simulation.n.2,       file = file.path(path.n,"wf.simulation.n.2.csv"),       row.names = FALSE)
write.csv(wf.simulation.n.20,      file = file.path(path.n,"wf.simulation.n.20.csv"),      row.names = FALSE)
write.csv(wf.simulation.n.200,     file = file.path(path.n,"wf.simulation.n.200.csv"),     row.names = FALSE)
write.csv(wf.simulation.n.2000,    file = file.path(path.n,"wf.simulation.n.2000.csv"),    row.names = FALSE)

path.n.l.p = "../data/WF_model/sim.n.low.p"
dir.create(path.n.l.p, showWarnings = FALSE)
write.csv(wf.simulation.n.20.p.1,  file = file.path(path.n.l.p,"wf.simulation.n.20.p.1.csv"),  row.names = FALSE)
write.csv(wf.simulation.n.20.p.3,  file = file.path(path.n.l.p,"wf.simulation.n.20.p.3.csv"),  row.names = FALSE)
write.csv(wf.simulation.n.20.p.5,  file = file.path(path.n.l.p,"wf.simulation.n.20.p.5.csv"),  row.names = FALSE)
write.csv(wf.simulation.n.20.p.9,  file = file.path(path.n.l.p,"wf.simulation.n.20.p.9.csv"),  row.names = FALSE)

path.n.h.p = "../data/WF_model/sim.n.hgh.p"
dir.create(path.n.h.p, showWarnings = FALSE)
write.csv(wf.simulation.n.1e4.p.1, file = file.path(path.n.h.p,"wf.simulation.n.1e4.p.1.csv"), row.names = FALSE)
write.csv(wf.simulation.n.1e4.p.3, file = file.path(path.n.h.p,"wf.simulation.n.1e4.p.3.csv"), row.names = FALSE)
write.csv(wf.simulation.n.1e4.p.5, file = file.path(path.n.h.p,"wf.simulation.n.1e4.p.5.csv"), row.names = FALSE)
write.csv(wf.simulation.n.1e4.p.9, file = file.path(path.n.h.p,"wf.simulation.n.1e4.p.9.csv"), row.names = FALSE)

write.csv(x = dat.hw, file = "../data/HW_principle/hw_simulation.csv", row.names = FALSE)
```


# References
